core/src/main/scala/com/raphtory/internals/management/QuerySender.scala

```diff
@@ -33,32 +33,47 @@ private[raphtory] class QuerySender(
 
   protected val logger: Logger = Logger(LoggerFactory.getLogger(this.getClass))
 
-  private val graphID               = config.getString("raphtory.graph.id")
-  val partitionServers: Int         = config.getInt("raphtory.partitions.serverCount")
-  val partitionsPerServer: Int      = config.getInt("raphtory.partitions.countPerServer")
-  val totalPartitions: Int          = partitionServers * partitionsPerServer
-  private lazy val writers          = topics.graphUpdates(graphID).endPoint
-  private lazy val queryManager     = topics.blockingIngestion.endPoint
-  private lazy val submissions      = topics.submissions.endPoint
-  private val blockingSources       = ArrayBuffer[Int]()
-  private var individualUpdateIndex = 0
-  private var lastQueryIndex        = 0
-  lazy val sourceID: Int            = getSourceID()
-
-  private def getSourceID(): Int =
-    idManager.getNextAvailableID() match {
-      case Some(id) =>
-        id
-      case None     =>
-        throw new NoIDException(s"Client '$clientID' was not able to acquire a source ID")
-    }
+  private val graphID                = config.getString("raphtory.graph.id")
+  val partitionServers: Int          = config.getInt("raphtory.partitions.serverCount")
+  val partitionsPerServer: Int       = config.getInt("raphtory.partitions.countPerServer")
+  val totalPartitions: Int           = partitionServers * partitionsPerServer
+  private lazy val writers           = topics.graphUpdates(graphID).endPoint
+  private lazy val queryManager      = topics.blockingIngestion.endPoint
+  private lazy val submissions       = topics.submissions.endPoint
+  private val blockingSources        = ArrayBuffer[Int]()
+  private var totalUpdateIndex       = 0     //used at the secondary index for the client
+  private var indexSinceLastIDChange = 0     //used to know how many messages to wait for when blocking in the Q manager
+  private var lastQueryIndex         = 0     //used to see if a prior query already sent out the unblocking message
+  private var newIDRequiredOnUpdate  = false // has a query been since since the last update and do I need a new ID
+  private var currentSourceID        = -1
+
+  def sourceID(update: Boolean): Int = {
+
+    if (
+            currentSourceID == -1 || (update && newIDRequiredOnUpdate)
+    ) //updates the sourceID if we haven't had one yet or if the user has sent a query since the last update block
+      idManager.getNextAvailableID() match {
+        case Some(id) =>
+          currentSourceID = id
+          newIDRequiredOnUpdate = false
+        case None     =>
+          throw new NoIDException(s"Client '$clientID' was not able to acquire a source ID")
+      }
+
+    currentSourceID
+  }
+
+  def getIndex: Int = totalUpdateIndex
 
   def submit(query: Query, customJobName: String = ""): QueryProgressTracker = {
 
-    if (lastQueryIndex < individualUpdateIndex) {
```

Comment:
could just check for `indexSinceLastIDChange > 0` and remove `lastQueryIndex` (one less thing to keep track of)